File: examples/go/escrow/Escrow.tsop.go
package contract

import "tsop"

type Escrow struct {
	tsop.SmartContract
	Buyer  tsop.PubKey `tsop:"readonly"`
	Seller tsop.PubKey `tsop:"readonly"`
	Arbiter tsop.PubKey `tsop:"readonly"`
}

func (c *Escrow) ReleaseBySeller(sig tsop.Sig) {
	tsop.Assert(tsop.CheckSig(sig, c.Seller))
}

func (c *Escrow) ReleaseByArbiter(sig tsop.Sig) {
	tsop.Assert(tsop.CheckSig(sig, c.Arbiter))
}

func (c *Escrow) RefundToBuyer(sig tsop.Sig) {
	tsop.Assert(tsop.CheckSig(sig, c.Buyer))
}

func (c *Escrow) RefundByArbiter(sig tsop.Sig) {
	tsop.Assert(tsop.CheckSig(sig, c.Arbiter))
}

File: examples/rust/escrow/Escrow.tsop.rs
use tsop::prelude::*;

#[tsop::contract]
pub struct Escrow {
    #[readonly]
    pub buyer: PubKey,
    #[readonly]
    pub seller: PubKey,
    #[readonly]
    pub arbiter: PubKey,
}

#[tsop::methods(Escrow)]
impl Escrow {
    #[public]
    pub fn release_by_seller(&self, sig: &Sig) {
        assert!(check_sig(sig, &self.seller));
    }

    #[public]
    pub fn release_by_arbiter(&self, sig: &Sig) {
        assert!(check_sig(sig, &self.arbiter));
    }

    #[public]
    pub fn refund_to_buyer(&self, sig: &Sig) {
        assert!(check_sig(sig, &self.buyer));
    }

    #[public]
    pub fn refund_by_arbiter(&self, sig: &Sig) {
        assert!(check_sig(sig, &self.arbiter));
    }
}
